### PR TITLE
Move testing to Github Actions

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py src/test/unit`
- [ ] Run cpplint: `make cpplint`
- [ ] Declare copyright holder and open-source license: see below

#### Summary

This PR will move tests from TravisCI to GithubActions.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
@storopoli mind reviewing this that it makes sense? I am a bit busy today/tomorrow and could use some help if you have a few moments.

**Comment by storopoli** [comment]
Oh, @serban-nicusor-toptal was faster! I will cancel the #3009 and review this instead...

**Comment by rok-cesnovar** [comment]
I commented in the other PR, I am fine whichever way.

**Review by storopoli — APPROVED** [review]
Nice work with the `matrix` for `os` and `python-version` foreseeing the future.

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        python-version: [env.pythonDefaultValue]
+      fail-fast: false
+    steps:
+      - name: Check out source code
+        uses: actions/checkout@v2
+        with:
+          submodules:  true
+      - name: Install clang++
+        run: sudo apt-get install clang-${{ env.clangppDefaultValue }}++
+      - name: Run tests
```
Comment (storopoli) [inline .github/workflows/testing.yml:34]: Is this necessary?

```bash
echo "CXX=clang-6.0" >> make/local
echo "O=0" >> make/local
```

**Comment by storopoli** [comment]
@serban-nicusor-toptal does by default the tests run with `-O0`?

File: .github/workflows/testing.yml (hunk 1 above)
Comment (serban-nicusor-toptal) [reply .github/workflows/testing.yml:35]: Looking at the log, it looks like not.
```
sudo apt-get install clang-6.0++
  shell: /bin/bash -e {0}
  env:
    pythonDefaultValue: 3.8
    clangppDefaultValue: 6.0
Reading package lists...
Building dependency tree...
Reading state information...
clang-6.0 is already the newest version (1:6.0-1ubuntu2).
The following NEW packages will be installed:
  clang-6.0-doc clang-6.0-examples libclang-6.0-dev python-clang-6.0
0 upgraded, 4 newly installed, 0 to remove and 13 not upgraded.
```
I'm not sure if we need/make use of any in `clang-6.0-doc clang-6.0-examples libclang-6.0-dev python-clang-6.0`

Comment (rok-cesnovar) [reply .github/workflows/testing.yml:35]: O=0 is there to make the tests run faster and eat up less RAM (Stan tests with models are known to take up a ton of RAM).

`CXX=clang-6.0` may not be necessary, we just need to convince tests to use clang 6 some way.

**Comment by serban-nicusor-toptal** [comment]
To be honest, I don't know. If it helps, our CI is only setting that up for `Math functions expressions test` but it's not used anywhere else.

**Comment by storopoli** [comment]
Ok, then I think it will be fine. Let's wait for the test to be finished

**Review by storopoli — APPROVED** [review]
Great work on the workflow. We can expand in the future with matrix.os and matrix.python-version.

**Comment by serban-nicusor-toptal** [comment]
What do you think about the `sudo apt-get install clang-6.0++` part ? If we don't use any of those 4 packages it becomes rendundant code.

**Comment by storopoli** [comment]
It seems that `ubuntu-latest` (which is right now `ubuntu-20-04` has already the [following installed softwares](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md):

```
...
Bash 5.0.17(1)-release
Clang 8.0.1, 9.0.1, 10.0.0
Erlang 11.1.7
GNU C++ 7.5.0, 8.4.0, 9.3.0, 10.2.0
GNU Fortran 8.4.0, 9.3.0, 10.2.0
...
```

I would leave it as it is, because it is future-proofed. What if GitHub Actions `ubuntu-latest` changes stuff, specially regarding Clang? We would get errors. If we leave `sudo apt-get install clang-6.0++` we probably would get a much more specific error and easier to handle and identify the cause also. And, one last thing, GitHub Actions are free for public repositories, so we don't have to worry that much with wasting resources.

File: .github/workflows/testing.yml [hunk 2]
```diff
@@ -33,6 +33,7 @@ jobs:
       - name: Install clang++
         run: sudo apt-get install clang-${{ env.clangppDefaultValue }}++
       - name: Run tests
-        env:
-          CXX: clang++-${{ env.clangppDefaultValue }}
-        run: python runTests.py -j 2 src/test/unit
+        run: |
+          echo "CXX=clang-${{ env.clangppDefaultValue }}" >> make/local
```
Comment (storopoli) [inline .github/workflows/testing.yml:37]: isn't better to specify

```bash
echo "CXX=clang++-${{ env.clangppDefaultValue }}" >> make/local
```

than

```bash
echo "CXX=clang-${{ env.clangppDefaultValue }}" >> make/local
```

Comment (serban-nicusor-toptal) [reply .github/workflows/testing.yml]: You're right, most probably a typo. Thanks!

**Comment by serban-nicusor-toptal** [comment]
We should be good for now, if there's anything else to add please let me know.
I've also removed `.travis-ci.yml` so after this PR is merged in, it won't run anymore.
If everything is alright please merge it or let me know to do it. Thanks!

**Review by rok-cesnovar — APPROVED** [review]
Thanks @serban-nicusor-toptal and @storopoli !

**Review by storopoli — APPROVED** [review]

**Pull request merged**